### PR TITLE
Add VQSR

### PR DESCRIPTION
**Pull request by vladsavelyev**
Add the [allele-specific VQSR pipeline](https://gatk.broadinstitute.org/hc/en-us/articles/360035890551-Allele-specific-annotation-and-filtering-of-germline-short-variants).

The GATK commands and their parameters are mostly borrowed from [WARP](https://github.com/broadinstitute/warp/blob/a9a3f549cac778badbbcb8d84865d80ee893752a/tasks/broad/JointGenotypingTasks.wdl).

I know there is a lot of code. But the code has been around for almost 2 years and been used to process all samples at the CPG, so should be well tested and doesn't need a thorough review. But if you spot any inefficiency or potential bugs, please point those out!

### CODE REVIEW

File: jobs/vqsr.py [hunk 1]
```diff
@@ -0,0 +1,1033 @@
...
+      --create-output-variant-index true \\
+      {f'-L {interval} ' if interval else ''} \\
+      {'--use-allele-specific-annotations ' if use_as_annotations else ''} \\
+      -mode INDEL
+      
+    df -h; pwd; du -sh $(dirname {j.output_vcf['vcf.gz']})
+
+    rm {input_vcf['vcf.gz']} {indels_recalibration} {indels_tranches}
+    rm -rf $TMP_DIR
+    mkdir $TMP_DIR
```
Comment (MattWellie) [inline jobs/vqsr.py:891]: just being pedantic, but wouldn't this save a command?
```suggestion
    rm -rf $TMP_DIR/*
```

File: jobs/vqsr.py [hunk 2]
```diff
@@ -0,0 +1,1033 @@
...
+      --tranches-file {indels_tranches} \\
+      --truth-sensitivity-filter-level {indel_filter_level} \\
+      --create-output-variant-index true \\
+      {f'-L {interval} ' if interval else ''} \\
+      {'--use-allele-specific-annotations ' if use_as_annotations else ''} \\
+      -mode INDEL
+      
+    df -h; pwd; du -sh $(dirname {j.output_vcf['vcf.gz']})
+
+    rm {input_vcf['vcf.gz']} {indels_recalibration} {indels_tranches}
```
Comment (MattWellie) [inline jobs/vqsr.py:889]: is this just to keep attached storage to a minimum?

File: jobs/vqsr.py [hunk 3]
```diff
@@ -0,0 +1,1033 @@
...
+    'AS_FS',
+    'AS_SOR',
+    # Not using depth for the following reasons:
+    # 1. The Broad pipelines don't use it;
+    # 2. -G AS_StandardAnnotation flag to GenotypeGVCFs doesn't include it;
+    # 3. For exomes, depth is an irrelevant feature and should be skipped.
+    # 'AS_VarDP',
+]
+SNP_ALLELE_SPECIFIC_FEATURES = ALLELE_SPECIFIC_FEATURES + ['AS_MQ']
+INDEL_ALLELE_SPECIFIC_FEATURES = ALLELE_SPECIFIC_FEATURES
```
Comment (MattWellie) [inline jobs/vqsr.py:49]: Nit
```suggestion
INDEL_ALLELE_SPECIFIC_FEATURES = [
    'AS_ReadPosRankSum',
    'AS_MQRankSum',
    'AS_QD',
    'AS_FS',
    'AS_SOR',
    # Not using depth for the following reasons:
    # 1. The Broad pipelines don't use it;
    # 2. -G AS_StandardAnnotation flag to GenotypeGVCFs doesn't include it;
    # 3. For exomes, depth is an irrelevant feature and should be skipped.
    # 'AS_VarDP',
]
SNP_ALLELE_SPECIFIC_FEATURES = INDEL_ALLELE_SPECIFIC_FEATURES + ['AS_MQ']
```

**Review by MattWellie — COMMENTED** [review]
LGTM, but I'm not super familiar with VQSR

File: jobs/vqsr.py (hunk 2 above)
Comment (vladsavelyev) [reply jobs/vqsr.py:889]: Yep, that's right!

File: jobs/vqsr.py [hunk 4]
```diff
@@ -0,0 +1,1033 @@
...
+                max_age='8h',
+                packages=[
+                    'cpg_utils',
+                    'cpg_gnomad',  # github.com/populationgenomics/gnomad_methods
+                    'click',
+                    'google',
+                    'fsspec',
+                    'sklearn',
+                    'gcloud',
+                    'selenium',
```
Comment (lgruen) [inline jobs/vqsr.py:177]: Do we need all these?

File: jobs/vqsr.py [hunk 5]
```diff
@@ -0,0 +1,1033 @@
...
+    # To make type checkers happy:
+    assert isinstance(indel_recalibrator_j.recalibration, hb.ResourceGroup)
+    assert isinstance(indel_recalibrator_j.tranches, hb.ResourceFile)
+    jobs.append(indel_recalibrator_j)
+
+    snp_max_gaussians = 6
+    if is_small_callset:
+        snp_max_gaussians = 4
+    elif is_huge_callset:
+        snp_max_gaussians = 8
```
Comment (lgruen) [inline jobs/vqsr.py:226]: What's the rationale for changing these depending on the size of the callset?

File: jobs/vqsr.py [hunk 6]
```diff
@@ -0,0 +1,1033 @@
...
+    # Not using depth for the following reasons:
+    # 1. The Broad pipelines don't use it;
+    # 2. -G AS_StandardAnnotation flag to GenotypeGVCFs doesn't include it;
+    # 3. For exomes, depth is an irrelevant feature and should be skipped.
+    # 'AS_VarDP',
+]
+SNP_ALLELE_SPECIFIC_FEATURES = ALLELE_SPECIFIC_FEATURES + ['AS_MQ']
+INDEL_ALLELE_SPECIFIC_FEATURES = ALLELE_SPECIFIC_FEATURES
+
+SNP_RECALIBRATION_TRANCHE_VALUES = [
```
Comment (lgruen) [inline jobs/vqsr.py:51]: Where do these values come from?

**Review by lgruen — APPROVED** [review]

File: jobs/vqsr.py (hunk 5 above)
Comment (vladsavelyev) [reply jobs/vqsr.py]: That was what they used in WARP before, but now it's just 6 consistently for SNPs, changed that accordingly

File: jobs/vqsr.py (hunk 4 above)
Comment (vladsavelyev) [reply jobs/vqsr.py]: We don't actually run VQSR from a matrix table in this repo (that's part of the large-cohort pipeline), so removed calling that script.

File: jobs/vqsr.py (hunk 6 above)
Comment (vladsavelyev) [reply jobs/vqsr.py:51]: Here and here
https://github.com/broadinstitute/warp/blob/79261cde9bd06bb6b1d4a83d75dc54f734541fec/pipelines/broad/dna_seq/germline/joint_genotyping/exome/JointGenotyping.inputs.json#L8-L11
https://github.com/broadinstitute/warp/blob/79261cde9bd06bb6b1d4a83d75dc54f734541fec/pipelines/broad/dna_seq/germline/joint_genotyping/wgs/JointGenotyping.inputs.json#L29-L35 (there is no direct example config for WGS AS-VQSR, but adjusted correspondingly)

Comment (vladsavelyev) [reply jobs/vqsr.py:57]: Added a comment into the docstring

**Pull request merged**